JS: Add query for detecting potential DOS form a tainted .length property

erik-krogh opened the pull request:
Add a new query that detects loops that might loop indefinitely or crash the server if given a creative user defined object.

The below provides a simple example: 

```JavaScript
var express = require('express');
var router = new express.Router();
var rootRoute = router.route('/foobar');
rootRoute.post(function(req, res) {
    var val = req.body;
    var ret = [];
    for (var i = 0; i < val.length; i++) {
        ret.push(val[i]);
    }
});
```

If an attacker gives an object of the form `{length:999999999}` as the `val`, then the server will keep looping until the `ret` array becomes too large and the server runs out of memory. 

It finds a good chunk of errors: https://lgtm.com/query/1254533150266280038/
I've managed eliminate most of the false positives. 
However, some of the query hits are false positives due to an exception being thrown within a few iterations of the loop by a method that is called within the loop body.

asger-semmle commented on javascript/ql/src/Security/CWE-834/TaintedLength.ql:
```diff
@@ -0,0 +1,22 @@
+ * @problem.severity warning
+ * @id js/tainted-length-looping
+ * @tags security
+ * @precision low
```
I think we can go all the way to `high` unless there are a lot of FPs. Otherwise `medium`.

erik-semmle replied:
I expect a decent amount of FPs, so I put it to as `medium` for now.

asger-semmle commented on javascript/ql/src/Security/CWE-834/examples/TaintedLength.js:
```diff
@@ -0,0 +1,22 @@
+rootRoute.post(function(req, res) {
+    var obj = req.body;
+
+    calcStuff(obj);
```
Since these examples will be embedded in the help page, please try to shorten them a bit:
- Remove the `'use strict'` directive.
- Inline `calcStuff`
- Use `var app = express();` and `app.post("/foo", (req, res) => ...)` instead of using routers.

asger-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLength.qll:
```diff
@@ -0,0 +1,43 @@
+   * A taint-tracking configuration for reasoning about looping on tainted objects with unbounded length.
+   */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "LabeledLoopTaintedObject" }
```
`this = "TaintedLength"`

asger-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,285 @@
+      exists(RelationalComparison compare, DataFlow::PropRead lengthRead |
+        compare = loop.getTest() and
+        compare.getLesserOperand() = indexVariable.getAnAccess() and
+        lengthRead.accesses(_, "length") and
```
`lengthRead.getPropertyName() = "length"`

asger-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,285 @@
+        compare = loop.getTest() and
+        compare.getLesserOperand() = indexVariable.getAnAccess() and
+        lengthRead.accesses(_, "length") and
+        lengthRead.flowsToExpr(compare.getGreaterOperand())
```
Could you move `lengthRead` into a field with a getter predicate so it can be used from `LoopSink`.

asger-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,285 @@
+  private class LoopSink extends Sink {
+    LoopSink() {
+      exists(ArrayIterationLoop loop, Expr lengthAccess, DataFlow::PropRead lengthRead |
+        loop.getTest().(RelationalComparison).getGreaterOperand() = lengthAccess and
```
... AFAICT this repeats the logic for matching the `length` access above, hence the suggestion about the field

erik-semmle replied:
It does, and yes moving lengthRead into a field fixes it.

asger-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,285 @@
+        lengthRead.accesses(this, "length") and
+        
+        // In the DOS we are looking for arrayRead will evaluate to undefined.
+        // If an obvious nullpointer happens on this undefined, then the DOS cannot happen.
```
> If an obvious nullpointer happens on this undefined

It is not really obvious what this means. Maybe just continue the line above with "this may cause an exception to be thrown, thus bailing out of the loop".

asger-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,256 @@
+  abstract class Sink extends DataFlow::Node { }
+
+  // for (..; .. sink.length; ...) ...
+  private class LoopSink extends Sink {
```
Sorry the qldoc I proposed was wrong. Could you mention that sink it the receiver of the length access?

erik-semmle commented on javascript/ql/src/Security/CWE-834/TaintedLength.qhelp:
```diff
@@ -7,32 +7,18 @@
         Iterating the elements of an untrusted object using the 
         <code>.length</code> property can lead to a server looping 
-        indefinitely or crashing. Thereby creating a 
+        indefinitely or crashing. This looping or crashing creates a 
```
By "crashing" i meant crashing the entire process (usually because out-of-mem). 

But i can see that it wasn't clear, so I'll focus on looping indefinitely.

asger-semmle commented on javascript/ql/src/Security/CWE-834/TaintedLength.qhelp:
```diff
@@ -7,11 +7,12 @@
+        indefinitely. This looping causes a denial-of-service or DoS by 
+        causing the server to hang or run out of memory. 
+        This happens when the server expects an array but an attacker sends 
+        a regular JSON object with an huge number in the 
```
an huge -> a huge

asger-semmle commented on javascript/ql/src/Security/CWE-834/TaintedLength.qhelp:
```diff
@@ -7,11 +7,12 @@
+        causing the server to hang or run out of memory. 
+        This happens when the server expects an array but an attacker sends 
+        a regular JSON object with an huge number in the 
+        <code>.length</code> property, such as `{length: 1e100}`, that the
```
`<code>` instead of backticks

xiemaisi commented on javascript/ql/src/Security/CWE-834/TaintedLength.qhelp:
```diff
@@ -0,0 +1,46 @@
+
+<overview>
+	<p>
+		Using the .length property of an untrusted object as a loop bound may
```
```suggestion
		Using the <code>.length</code> property of an untrusted object as a loop bound may
```

(for consistency)

xiemaisi commented on javascript/ql/src/Security/CWE-834/TaintedLength.qhelp:
```diff
@@ -0,0 +1,46 @@
+		cause indefinite looping since a malicious attacker can set the
+		<code>.length</code> property to a very large number. For example, 
+		when a program that expects an array is passed a JSON object such as
+		<code>{length: 1e100}</code>, the loop will be run for 1e100 
```
I'd slightly prefer "10<sup>100</sup>" over "1e100" for clarity. Unless you have a particular reason to prefer the latter?

erik-semmle replied:
I didn't know <sup> was a thing I could use. I changed it.

xiemaisi commented on javascript/ql/src/Security/CWE-834/TaintedLength.ql:
```diff
@@ -0,0 +1,20 @@
+/**
+ * @name Tainted .length in loop condition
+ * @description Iterating over an object with a user-controlled .length
+ *              property can cause indefinite looping
```
```suggestion
 *              property can cause indefinite looping.
```

xiemaisi commented on javascript/ql/src/Security/CWE-834/examples/TaintedLength_fixed.js:
```diff
@@ -0,0 +1,16 @@
+app.post("/foo", (req, res) => {
+    var obj = req.body;
+    
+    if (!(obj instanceof Array)) { // prevents DOS
```
```suggestion
    if (!(obj instanceof Array)) { // prevents DoS
```

xiemaisi commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,281 @@
+          arrayRead.flowsToExpr(throws) and
+          isCrashingWithNullValues(throws)
+        ) and
+        // The existence of some kind of early-exit usually indicates that the loop will stop early and no DOS happens.
```
```suggestion
        // The existence of some kind of early-exit usually indicates that the loop will stop early and no DoS happens.
```

xiemaisi commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  /**
+   * Holds if `name` is a method from lodash vulnerable to a DOS attack if called with a tained object.
```
```suggestion
   * Holds if `name` is a method from lodash vulnerable to a DoS attack if called with a tainted object.
```

xiemaisi commented on javascript/ql/src/semmle/javascript/security/dataflow/TaintedLengthCustomizations.qll:
```diff
@@ -0,0 +1,281 @@
+              isCrashingWithNullValues(throws)
+            )
+            or
+            // similar to the loop sink - the existence of an early-exit usually means that no DOS can happen.
```
```suggestion
            // similar to the loop sink - the existence of an early-exit usually means that no DoS can happen.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitBad.js:
```diff
@@ -0,0 +1,69 @@
+	for (var i = 0; i < val.length; i++) { // NOT OK!
+	  for (var k = 0; k < 2; k++) {
+		  if (k == 3) {
+		      // Does not prevent DOS, because this is inside an inner loop.
```
```suggestion
		      // Does not prevent DoS, because this is inside an inner loop.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitBad.js:
```diff
@@ -0,0 +1,69 @@
+	for (var i = 0; i < val.length; i++) { // NOT OK!
+	  if (val[i] == null) {
+		  try {
+			  throw 2; // Is catched, and therefore the DOS is not prevented.
```
```suggestion
			  throw 2; // Is caught, and therefore the DoS is not prevented.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitBad.js:
```diff
@@ -0,0 +1,69 @@
+	for (var i = 0; i < val.length; i++) { // NOT OK!
+	  if (val[i] == null) {
+		  (function (i) {
+			  return i+2; // Does not prevent DOS.
```
```suggestion
			  return i+2; // Does not prevent DoS.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitBad.js:
```diff
@@ -0,0 +1,69 @@
+	_.map(val, function (e) {
+		if (!e) {
+			try {
+				throw new Error(); // Does not prevent DOS
```
```suggestion
				throw new Error(); // Does not prevent DoS
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitGood.js:
```diff
@@ -0,0 +1,57 @@
+	
+	for (var i = 0; i < val.length; i++) { // OK
+	  if (val[i] == null) {
+		  break; // prevents DOS.
```
```suggestion
		  break; // prevents DoS.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitGood.js:
```diff
@@ -0,0 +1,57 @@
+	
+	for (var i = 0; i < val.length; i++) { // OK
+	  if (val[i] == null) {
+		  throw 2; // prevents DOS.
```
```suggestion
		  throw 2; // prevents DoS.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitGood.js:
```diff
@@ -0,0 +1,57 @@
+	
+	for (var i = 0; i < val.length; i++) { // OK
+	  if (val[i] == null) {
+		  return 2; // prevents DOS.
```
```suggestion
		  return 2; // prevents DoS.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthExitGood.js:
```diff
@@ -0,0 +1,57 @@
+function lodashThrow(val) {
+	_.map(val, function (e) { // OK
+		if (!e) {
+			throw new Error(); // prevents DOS.
```
```suggestion
			throw new Error(); // prevents DoS.
```

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-834/TaintedLengthObviousLengthCheck.js:
```diff
@@ -0,0 +1,22 @@
+function problem(val) {
+    var ret = [];
+
+    // Prevents DOS
```
```suggestion
    // Prevents DoS
```

xiemaisi commented on change-notes/1.23/analysis-javascript.md:
```diff
@@ -14,7 +14,7 @@
 | **Query**                                                                 | **Tags**                                                          | **Purpose**                                                                                                                                                                            |
 |---------------------------------------------------------------------------|-------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | Unused index variable (`js/unused-index-variable`)                        | correctness                                                       | Highlights loops that iterate over an array, but do not use the index variable to access array elements, indicating a possible typo or logic error. |
```
```suggestion
| Unused index variable (`js/unused-index-variable`)                        | correctness                                                       | Highlights loops that iterate over an array, but do not use the index variable to access array elements, indicating a possible typo or logic error. Results are shown on LGTM by default. |
```

(Not technically part of this PR, but might as well fix it.)

xiemaisi commented on change-notes/1.23/analysis-javascript.md:
```diff
@@ -14,7 +14,7 @@
 |---------------------------------------------------------------------------|-------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | Unused index variable (`js/unused-index-variable`)                        | correctness                                                       | Highlights loops that iterate over an array, but do not use the index variable to access array elements, indicating a possible typo or logic error. |
-
+| Tainted .length in loop condition (`js/loop-bound-injection`)             | security                                                          | Highlights loops where a user-controlled object with an arbitrary .length value can trick the server to loop infinitely. |
```
```suggestion
| Tainted .length in loop condition (`js/loop-bound-injection`)             | security                                                          | Highlights loops where a user-controlled object with an arbitrary .length value can trick the server to loop infinitely. Results are not shown on LGTM by default. |
```

erik-semmle commented:
I changed the name of all the files from `TaintedLength` to `LoopBoundInjection`

erik-semmle commented on javascript/ql/src/Security/CWE-834/TaintedLength.ql:
```diff
@@ -0,0 +1,22 @@
+ *              to loop infinitely. 
+ * @kind path-problem
+ * @problem.severity warning
+ * @id js/tainted-length-looping
```
I changed the ID and the name to match loop-bound-injection.

erik-semmle commented on javascript/ql/test/query-tests/Security/CWE-834/LoopBoundInjectionExitGood.js:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+// the obvious null-pointer detection should not hit this one.
```
I think you capitalized the wrong word.

shati-patel replied:
You're absolutely right - sorry about that! I'll edit.

erik-krogh commented:
I did some further performance evaluation, and my conclusion is `@kind path-problem` in JavaScript is generally slow, and that my query is fine. 

I did another dist-compare where i replaced the `where` section of the query with `where none()`. 
All the results where ±15% with/without `where none()`.

I took one of the slowest projects and removed all my code except for 
```
import semmle.javascript.security.dataflow.RemoteFlowSources
import semmle.javascript.security.TaintedObject
import DataFlow::PathGraph
```
And performance was basically identical to the full query.

So it seems to me that the `nodes` and `edges` queries from PathGraph are to blame, and not my query.

asger-semmle commented:
Sorry about the `where none()` suggestion - it turns out it doesn't actually prevent the query from running since `nodes` and `edges` force evaluation of the underlying taint-tracking problem anyway. We'll still want to evaluate the query in two revisions so we have something to compare against.

xiemaisi commented:
I ran the query locally on snapshots of Node.js and gecko, two of our heavy hitters, and didn't notice anything particularly slow in the query stage, so I think performance is probably fine.

erik-krogh commented on javascript/ql/src/semmle/javascript/security/dataflow/LoopBoundInjectionCustomizations.qll:
```diff
@@ -0,0 +1,281 @@
+module LoopBoundInjection {
+  import semmle.javascript.security.dataflow.RemoteFlowSources
+  import semmle.javascript.security.TaintedObject
+  import DataFlow::PathGraph
```
`semmle.javascript.security.TaintedObject` is to construct the sources.   
`DataFlow::PathGraph` is needed for the paths in LGTM, but maybe the import should be moved to `LoopBoundInjection.ql`?

We do not need `semmle.javascript.security.dataflow.RemoteFlowSources`, so I've removed that one.

xiemaisi replied:
Yeah, I was asking about `PathGraph`. that should be imported from the `.ql` file.

xiemaisi commented:
@asger-semmle, anything else you think needs addressing?

asger-semmle commented on change-notes/1.23/analysis-javascript.md:
```diff
@@ -13,8 +13,8 @@
-| Unused index variable (`js/unused-index-variable`)                        | correctness                                                       | Highlights loops that iterate over an array, but do not use the index variable to access array elements, indicating a possible typo or logic error. |
-
+| Unused index variable (`js/unused-index-variable`)                        | correctness                                                       | Highlights loops that iterate over an array, but do not use the index variable to access array elements, indicating a possible typo or logic error. Results are shown on LGTM by default. |
+| Tainted .length in loop condition (`js/loop-bound-injection`)             | security, external/cwe/cwe-834                                      | Highlights loops where a user-controlled object with an arbitrary .length value can trick the server to loop indefinitely. Results are not shown on LGTM by default. |
```
Query name needs an update

xiemaisi commented:
@asger-semmle, is there anything else you want @erik-krogh to address?
